### PR TITLE
Feature/16034 rollback query in wazuhdb

### PR DESCRIPTION
**Pull request by thejbte**
|Related issue|Manual Testing|
|---|---|
|https://github.com/wazuh/wazuh/issues/16034|[#3874](https://github.com/wazuh/wazuh-qa/issues/3874)|

## Description
This PR solves the issue https://github.com/wazuh/wazuh/issues/16034, the proposed solution was add a mechanism, called ROLLBACK, is something that needs to be implemented within WazuhDB to prevent these scenarios, which lead to corrupted databases.

### How testing:
1. Run wazuh server
2. open console and send the next command
    ```
    python3 wdb-query.py "global sql INSERT INTO agent (id, name, ip, register_ip, internal_key, date_add, 'group') VALUES (997,\"julian\",\"192.168.10.127\",\"192.168.10.127\",\"ABCDE\",1234,\"default\");"
    ```
     <details><summary>output</summary>

      ```json
      []

     ```
     </details>
3.  validate if query is in memory
     ```
     python3 wdb-query.py "global sql SELECT * FROM agent where id=997;"
     ```
     <details><summary>output</summary>

      ```json
      [
          {
              "id": 997,
              "name": "julian",
              "ip": "192.168.10.127",
              "register_ip": "192.168.10.127",
              "internal_key": "ABCDE",
              "node_name": "unknown",
              "date_add": 1234,
              "group": "default",
              "group_sync_status": "synced",
              "sync_status": "synced",
              "connection_status": "never_connected",
              "disconnection_time": 0,
              "group_config_status": "not synced"
          }
      ]

     ```
     </details>
4. validate if that query  doesn't commit yet
    ```
    sqlite3 /var/ossec/queue/db/global.db "SELECT * FROM agent where id=997;"
    ```
     <details><summary>output</summary>

      ```json


     ```
     </details>
5. Do the `ROLLBACK`
    ```
    python3 wdb-query.py "global rollback"
    ```
     <details><summary>output</summary>

      ```vim
      ok

     ```
     </details>
6. repeat again step `3` and `4`
     <details><summary>output 3</summary>

      ```json
      []

     ```
     </details>
     <details><summary>output 4</summary>

      ```json


     ```
     </details>

After doing an `insert` (or any other) to the database, we need to see what was kept in memory, then validate that the commit hasn't been done to the database yet, then the `rollback` needs to be done. and validate that what was in memory and has not been committed has not been added to the database either.

*Note:* When the transaction is done, the commit hasn't been done yet, these are done about every 10 seconds, so the above steps should be done as fast as possible as `rollback` only works for uncommitted transactions.

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation
- [x] Package installation
- [x] Source upgrade
- [x] Package upgrade
- [x] Review logs syntax and correct language

### CODE REVIEW

**Comment by thejbte** [comment]
### Valgrind

<details><summary>Valgrind report </summary>

```vim
========================

root@julian-A15-FA506QM:/home/thejbte/Documents# kill -9 $(pidof wazuh-db)
root@julian-A15-FA506QM:/home/thejbte/Documents# valgrind --leak-check=full --track-fds=yes /var/ossec/bin/wazuh-db -f
==198645== Memcheck, a memory error detector
==198645== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==198645== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==198645== Command: /var/ossec/bin/wazuh-db -f
==198645== 
2023/01/31 15:57:17 wazuh-db[198645] debug_op.c:70 at _log(): DEBUG: Logging module auto-initialized
2023/01/31 15:57:17 wazuh-db[198645] main.c:119 at main(): DEBUG: Wazuh home directory: /var/ossec
2023/01/31 15:57:17 wazuh-db[198645] main.c:140 at main(): DEBUG: Template file removed: /var/ossec/queue/db/.template.db
2023/01/31 15:57:17 wazuh-db[198645] main.c:146 at main(): ERROR: Could not set resource limit for file descriptors to 458752: Operation not permitted (1)
2023/01/31 15:57:17 wazuh-db[198645] main.c:196 at main(): INFO: Started (pid: 198645).
2023/01/31 15:57:17 wazuh-db[198645] main.c:454 at run_backup(): DEBUG: Database backup thread started.
2023/01/31 15:57:20 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:57:20 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:57:20 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:57:22 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:57:22 wazuh-db[198645] wdb_parser.c:1340 at wdb_parse(): DEBUG: Task query: upgrade_cancel_tasks {"node":"node01"}
2023/01/31 15:57:22 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:57:30 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:57:30 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:57:30 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:57:33 wazuh-db[198645] wdb.c:1075 at wdb_commit_old(): DEBUG: Agent 'tasks' database commited. Time: 3.963 ms.
2023/01/31 15:57:40 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:57:40 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:57:40 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:57:50 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:57:50 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:57:50 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:58:00 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:58:00 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:58:00 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:58:10 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:58:10 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:58:10 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:58:20 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:58:20 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:58:20 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:58:21 wazuh-db[198645] wdb.c:1075 at wdb_commit_old(): DEBUG: Agent 'global' database commited. Time: 0.071 ms.
2023/01/31 15:58:30 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:58:30 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:58:30 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:58:40 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:58:40 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:58:40 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:58:47 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:58:47 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: sql INSERT INTO agent (id, name, ip, register_ip, internal_key, date_add, 'group') VALUES (996,"julian","192.168.10.127","192.168.10.127","ABCDE",1234,"default");
2023/01/31 15:58:48 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:58:50 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:58:50 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:58:50 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:58:54 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:58:54 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: sql SELECT * FROM agent where id=996
2023/01/31 15:58:54 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:58:59 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:58:59 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: rollback
2023/01/31 15:58:59 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:59:00 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:59:00 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:59:00 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:59:10 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:59:10 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:59:10 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:59:13 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:59:13 wazuh-db[198645] wdb_parser.c:263 at wdb_parse(): DEBUG: Agent 000 query: rootcheck save 1675198753 Ending rootcheck scan.
2023/01/31 15:59:13 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:59:20 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:59:20 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:59:20 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:59:27 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:59:27 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: sql SELECT * FROM agent
2023/01/31 15:59:27 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:59:30 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:59:30 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:59:30 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:59:40 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:59:40 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:59:40 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 15:59:50 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 15:59:50 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 15:59:50 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 16:00:00 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 16:00:00 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 16:00:00 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 16:00:01 wazuh-db[198645] wdb.c:1075 at wdb_commit_old(): DEBUG: Agent 'global' database commited. Time: 0.083 ms.
2023/01/31 16:00:10 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 16:00:10 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 16:00:10 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 16:00:20 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 16:00:20 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 16:00:20 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 16:00:30 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 16:00:30 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 16:00:30 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 16:00:40 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 16:00:40 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 16:00:40 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
2023/01/31 16:00:50 wazuh-db[198645] main.c:324 at run_dealer(): DEBUG: New client connected (5).
2023/01/31 16:00:50 wazuh-db[198645] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 16:00:50 wazuh-db[198645] main.c:386 at run_worker(): DEBUG: Client 5 disconnected.
^C2023/01/31 16:00:55 wazuh-db[198645] main.c:558 at handler(): INFO: (1225): SIGNAL [(2)-(Interrupt)] Received. Exit Cleaning...
2023/01/31 16:00:56 wazuh-db[198645] wdb.c:1029 at wdb_close_all(): DEBUG: Closing all databases...
2023/01/31 16:00:56 wazuh-db[198645] wdb.c:1033 at wdb_close_all(): DEBUG: Closing database for agent 000
2023/01/31 16:00:56 wazuh-db[198645] wdb.c:1033 at wdb_close_all(): DEBUG: Closing database for agent 001
2023/01/31 16:00:56 wazuh-db[198645] wdb.c:1033 at wdb_close_all(): DEBUG: Closing database for agent global
2023/01/31 16:00:56 wazuh-db[198645] wdb.c:1033 at wdb_close_all(): DEBUG: Closing database for agent tasks
2023/01/31 16:00:56 wazuh-db[198645] main.c:265 at main(): DEBUG: Template file removed again: queue/db/.template.db
==198645== 
==198645== FILE DESCRIPTORS: 4 open (3 std) at exit.
==198645== Open file descriptor 3: /dev/urandom
==198645==    at 0x52A46EB: open (open64.c:41)
==198645==    by 0x1DA442: randombytes (in /var/ossec/bin/wazuh-db)
==198645==    by 0x1DA506: os_random (in /var/ossec/bin/wazuh-db)
==198645==    by 0x1D4163: OSHash_Create (in /var/ossec/bin/wazuh-db)
==198645==    by 0x11A37C: main (in /var/ossec/bin/wazuh-db)
==198645== 
==198645== 
==198645== HEAP SUMMARY:
==198645==     in use at exit: 0 bytes in 0 blocks
==198645==   total heap usage: 14,474 allocs, 14,474 frees, 43,299,925 bytes allocated
==198645== 
==198645== All heap blocks were freed -- no leaks are possible
==198645== 
==198645== For lists of detected and suppressed errors, rerun with: -s
==198645== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
==198645== could not unlink /tmp/vgdb-pipe-from-vgdb-to-198645-by-root-on-???
==198645== could not unlink /tmp/vgdb-pipe-to-vgdb-from-198645-by-root-on-???
==198645== could not unlink /tmp/vgdb-pipe-shared-mem-vgdb-198645-by-root-on-???
```
</details>

**Comment by thejbte** [comment]
### ThreadSanitizer

<details><summary>ThreadSanitizer report </summary>

```vim
root@julian-A15-FA506QM:/home/thejbte/Documents# /var/ossec/bin/wazuh-db -f -u root
/var/ossec/bin/wazuh-db: invalid option -- 'u'
 
Wazuh v4.4.0 - Wazuh Inc. (info@wazuh.com)
http://www.wazuh.com
  wazuh-db: -[Vhdtf]
    -V          Version and license message.
    -h          This help message.
    -d          Debug mode. Use this parameter multiple times to increase the debug level.
    -t          Test configuration.
    -f          Run in foreground.
root@julian-A15-FA506QM:/home/thejbte/Documents# /var/ossec/bin/wazuh-db -f
2023/01/31 17:11:51 wazuh-db[214888] debug_op.c:70 at _log(): DEBUG: Logging module auto-initialized
2023/01/31 17:11:51 wazuh-db[214888] main.c:119 at main(): DEBUG: Wazuh home directory: /var/ossec
2023/01/31 17:11:51 wazuh-db[214888] main.c:140 at main(): DEBUG: Template file removed: /var/ossec/queue/db/.template.db
2023/01/31 17:11:51 wazuh-db[214888] main.c:196 at main(): INFO: Started (pid: 214888).
2023/01/31 17:11:51 wazuh-db[214888] main.c:454 at run_backup(): DEBUG: Database backup thread started.
2023/01/31 17:11:52 wazuh-db[214888] main.c:324 at run_dealer(): DEBUG: New client connected (7).
2023/01/31 17:11:52 wazuh-db[214888] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 17:11:52 wazuh-db[214888] main.c:386 at run_worker(): DEBUG: Client 7 disconnected.
2023/01/31 17:12:02 wazuh-db[214888] main.c:324 at run_dealer(): DEBUG: New client connected (6).
2023/01/31 17:12:02 wazuh-db[214888] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 17:12:02 wazuh-db[214888] main.c:386 at run_worker(): DEBUG: Client 6 disconnected.
2023/01/31 17:12:09 wazuh-db[214888] main.c:324 at run_dealer(): DEBUG: New client connected (6).
2023/01/31 17:12:09 wazuh-db[214888] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: sql SELECT * FROM agent where id=997
2023/01/31 17:12:09 wazuh-db[214888] main.c:386 at run_worker(): DEBUG: Client 6 disconnected.
2023/01/31 17:12:12 wazuh-db[214888] main.c:324 at run_dealer(): DEBUG: New client connected (6).
2023/01/31 17:12:12 wazuh-db[214888] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 17:12:12 wazuh-db[214888] main.c:386 at run_worker(): DEBUG: Client 6 disconnected.
2023/01/31 17:12:20 wazuh-db[214888] main.c:324 at run_dealer(): DEBUG: New client connected (6).
2023/01/31 17:12:20 wazuh-db[214888] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: sql INSERT INTO agent (id, name, ip, register_ip, internal_key, date_add, 'group') VALUES (997,"julian","192.168.10.127","192.168.10.127","ABCDE",1234,"default");
2023/01/31 17:12:20 wazuh-db[214888] main.c:386 at run_worker(): DEBUG: Client 6 disconnected.
2023/01/31 17:12:22 wazuh-db[214888] main.c:324 at run_dealer(): DEBUG: New client connected (6).
2023/01/31 17:12:22 wazuh-db[214888] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 17:12:22 wazuh-db[214888] main.c:386 at run_worker(): DEBUG: Client 6 disconnected.
2023/01/31 17:12:27 wazuh-db[214888] main.c:324 at run_dealer(): DEBUG: New client connected (6).
2023/01/31 17:12:27 wazuh-db[214888] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: sql SELECT * FROM agent where id=997
2023/01/31 17:12:27 wazuh-db[214888] main.c:386 at run_worker(): DEBUG: Client 6 disconnected.
2023/01/31 17:12:32 wazuh-db[214888] main.c:324 at run_dealer(): DEBUG: New client connected (6).
2023/01/31 17:12:32 wazuh-db[214888] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 17:12:32 wazuh-db[214888] main.c:386 at run_worker(): DEBUG: Client 6 disconnected.
2023/01/31 17:12:41 wazuh-db[214888] main.c:324 at run_dealer(): DEBUG: New client connected (6).
2023/01/31 17:12:41 wazuh-db[214888] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: sql SELECT * FROM agent
2023/01/31 17:12:41 wazuh-db[214888] main.c:386 at run_worker(): DEBUG: Client 6 disconnected.
2023/01/31 17:12:42 wazuh-db[214888] main.c:324 at run_dealer(): DEBUG: New client connected (6).
2023/01/31 17:12:42 wazuh-db[214888] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 17:12:42 wazuh-db[214888] main.c:386 at run_worker(): DEBUG: Client 6 disconnected.
2023/01/31 17:12:50 wazuh-db[214888] main.c:324 at run_dealer(): DEBUG: New client connected (6).
2023/01/31 17:12:50 wazuh-db[214888] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: rollback
2023/01/31 17:12:50 wazuh-db[214888] main.c:386 at run_worker(): DEBUG: Client 6 disconnected.
2023/01/31 17:12:52 wazuh-db[214888] main.c:324 at run_dealer(): DEBUG: New client connected (6).
2023/01/31 17:12:52 wazuh-db[214888] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 17:12:52 wazuh-db[214888] main.c:386 at run_worker(): DEBUG: Client 6 disconnected.
2023/01/31 17:12:52 wazuh-db[214888] main.c:324 at run_dealer(): DEBUG: New client connected (6).
2023/01/31 17:12:52 wazuh-db[214888] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: sql SELECT * FROM agent
2023/01/31 17:12:52 wazuh-db[214888] main.c:386 at run_worker(): DEBUG: Client 6 disconnected.
2023/01/31 17:13:02 wazuh-db[214888] main.c:324 at run_dealer(): DEBUG: New client connected (6).
2023/01/31 17:13:02 wazuh-db[214888] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 17:13:02 wazuh-db[214888] main.c:386 at run_worker(): DEBUG: Client 6 disconnected.
2023/01/31 17:13:12 wazuh-db[214888] main.c:324 at run_dealer(): DEBUG: New client connected (6).
2023/01/31 17:13:12 wazuh-db[214888] wdb_parser.c:821 at wdb_parse(): DEBUG: Global query: get-distinct-groups 
2023/01/31 17:13:12 wazuh-db[214888] main.c:386 at run_worker(): DEBUG: Client 6 disconnected.
^C==================
WARNING: ThreadSanitizer: signal-unsafe call inside of a signal (pid=214888)
    #0 <null> <null> (libtsan.so.0+0x31edc)
    #1 <null> <null> (wazuh-db+0x1438b9)
    #2 <null> <null> (wazuh-db+0x15fa0)

SUMMARY: ThreadSanitizer: signal-unsafe call inside of a signal (/lib/x86_64-linux-gnu/libtsan.so.0+0x31edc) 
==================
==================
WARNING: ThreadSanitizer: signal-unsafe call inside of a signal (pid=214888)
    #0 <null> <null> (libtsan.so.0+0x31c57)
    #1 <null> <null> (libc.so.6+0x7f6cd)
    #2 <null> <null> (wazuh-db+0x123504)
    #3 <null> <null> (wazuh-db+0x15fa0)

SUMMARY: ThreadSanitizer: signal-unsafe call inside of a signal (/lib/x86_64-linux-gnu/libtsan.so.0+0x31c57) 
==================
==================
WARNING: ThreadSanitizer: signal-unsafe call inside of a signal (pid=214888)
    #0 <null> <null> (libtsan.so.0+0x31c57)
    #1 <null> <null> (libc.so.6+0x7ec23)
    #2 <null> <null> (libtsan.so.0+0x441b2)
    #3 <null> <null> (wazuh-db+0x123626)
    #4 <null> <null> (wazuh-db+0x15fa0)

SUMMARY: ThreadSanitizer: signal-unsafe call inside of a signal (/lib/x86_64-linux-gnu/libtsan.so.0+0x31c57) 
==================
==================
WARNING: ThreadSanitizer: signal-unsafe call inside of a signal (pid=214888)
    #0 <null> <null> (libtsan.so.0+0x37ab8)
    #1 <null> <null> (libc.so.6+0x8dcf8)
    #2 <null> <null> (wazuh-db+0x12377b)
    #3 <null> <null> (wazuh-db+0x15fa0)

SUMMARY: ThreadSanitizer: signal-unsafe call inside of a signal (/lib/x86_64-linux-gnu/libtsan.so.0+0x37ab8) 
==================
==================
WARNING: ThreadSanitizer: signal-unsafe call inside of a signal (pid=214888)
    #0 <null> <null> (libtsan.so.0+0x37ab8)
    #1 <null> <null> (libc.so.6+0x7eda6)
    #2 <null> <null> (wazuh-db+0x12377b)
    #3 <null> <null> (wazuh-db+0x15fa0)

SUMMARY: ThreadSanitizer: signal-unsafe call inside of a signal (/lib/x86_64-linux-gnu/libtsan.so.0+0x37ab8) 
==================
2023/01/31 17:13:14 wazuh-db[214888] main.c:558 at handler(): INFO: (1225): SIGNAL [(2)-(Interrupt)] Received. Exit Cleaning...
==================
WARNING: ThreadSanitizer: signal-unsafe call inside of a signal (pid=214888)
    #0 <null> <null> (libtsan.so.0+0x37ab8)
    #1 <null> <null> (wazuh-db+0x1238ab)
    #2 <null> <null> (wazuh-db+0x15fa0)

SUMMARY: ThreadSanitizer: signal-unsafe call inside of a signal (/lib/x86_64-linux-gnu/libtsan.so.0+0x37ab8) 
==================
==================
WARNING: ThreadSanitizer: signal handler spoils errno (pid=214888)
    #0 <null> <null> (wazuh-db+0x1717d)
    #1 <null> <null> (wazuh-db+0x15fa0)

SUMMARY: ThreadSanitizer: signal handler spoils errno (/var/ossec/bin/wazuh-db+0x1717d) 
==================
2023/01/31 17:13:15 wazuh-db[214888] wdb.c:1029 at wdb_close_all(): DEBUG: Closing all databases...
2023/01/31 17:13:15 wazuh-db[214888] wdb.c:1033 at wdb_close_all(): DEBUG: Closing database for agent 000
2023/01/31 17:13:15 wazuh-db[214888] wdb.c:1033 at wdb_close_all(): DEBUG: Closing database for agent global
2023/01/31 17:13:15 wazuh-db[214888] wdb.c:1033 at wdb_close_all(): DEBUG: Closing database for agent 001
2023/01/31 17:13:15 wazuh-db[214888] main.c:265 at main(): DEBUG: Template file removed again: queue/db/.template.db
ThreadSanitizer: reported 7 warnings
```
</details>

File: src/wazuh_db/wdb.c [hunk 1]
```diff
@@ -1931,3 +1893,43 @@ bool wdb_check_backup_enabled() {
...
+
+    wdb->transaction = state;
+    if (1 == state) {
+        wdb->transaction_begin_time = time(NULL);
+    } else {
+        /* do nothing */
+    }
+
+    return 0;
+}
```
Comment (Nicogp) [inline src/wazuh_db/wdb.c:1935]: Add newline at the end

File: src/wazuh_db/wdb.c [hunk 2]
```diff
@@ -1931,3 +1893,43 @@ bool wdb_check_backup_enabled() {
...
+    if (wdb_ptr_any_txn != NULL) {
+        if (wdb_ptr_any_txn(wdb->db) == -1) {
+            return -1;
+        }
+    }
+
+    wdb->transaction = state;
+    if (1 == state) {
+        wdb->transaction_begin_time = time(NULL);
+    } else {
```
Comment (Nicogp) [inline src/wazuh_db/wdb.c:1930]: Remove the else block if not needed

File: src/wazuh_db/wdb.h [hunk 3]
```diff
@@ -707,6 +710,9 @@ int wdb_begin2(wdb_t * wdb);
 int wdb_commit(sqlite3 *db);
 int wdb_commit2(wdb_t * wdb);
 
+int wdb_rollback(sqlite3 *db);
+int wdb_rollback2(wdb_t * wdb);
```
Comment (Nicogp) [inline src/wazuh_db/wdb.h:714]: Add comments to `wdb_rollback` and `wdb_rollback2` functions

File: src/wazuh_db/wdb.h [hunk 4]
```diff
@@ -484,6 +484,9 @@ struct kv_list {
     const struct kv_list *next;
 };
 
+/** @brief pointer to function for any transaction*/
```
Comment (Nicogp) [inline src/wazuh_db/wdb.h:487]:
```
/**
 * @brief pointer to function for any transaction
 */
```

**Comment by thejbte** [comment]
- Manual rollback from API
api_wdb_test_rollback
##
- Automatic rollback from API
automatic_rollback_Peek 2023-03-24 16-59

**Comment by thejbte** [comment]
- Automatic rollback without API
rollback_auto

File: src/wazuh_db/wdb_sca.c [hunk 5]
```diff
@@ -493,7 +493,7 @@ int wdb_sca_policy_info_save(wdb_t * wdb,char *name,char * file,char * id,char *
     sqlite3_bind_text(stmt, 5, references, -1, NULL);
     sqlite3_bind_text(stmt, 6, hash_file, -1, NULL);
 
-    if (sqlite3_step(stmt) == SQLITE_DONE) {
+    if (wdb_step(stmt) == SQLITE_DONE) {
```
Comment (TomasTurina) [inline src/wazuh_db/wdb_sca.c:496]: This query modifies the DB.

File: src/wazuh_db/wdb_task.c
Comment (TomasTurina) [inline src/wazuh_db/wdb_task.c:1]: There is a query in `wdb_task_delete_old_entries` that modifies the DB.

**Review by TomasTurina — CHANGES_REQUESTED** [review]
Investigate if you can use `wdb_exec_stmt_silent` everywhere a rollback is required.

File: src/wazuh_db/wdb_upgrade.c [hunk 6]
```diff
@@ -340,7 +340,7 @@ int wdb_adjust_v4(wdb_t *wdb) {
         sqlite3_bind_text(update_stmt, 1, decoded_attrs, -1, NULL);
         sqlite3_bind_text(update_stmt, 2, file, -1, NULL);
 
-        if (sqlite3_step(update_stmt) != SQLITE_DONE) {
+        if (wdb_step1(update_stmt, wdb, WDB_NO_ATTEMPTS, true) != SQLITE_DONE) {
```
Comment (TomasTurina) [inline src/wazuh_db/wdb_upgrade.c:343]: I think it's not necessary to do rollback here since the error is not critical.

**Review by TomasTurina — CHANGES_REQUESTED** [review]

**Review by TomasTurina — CHANGES_REQUESTED** [review]
We need to define in which cases the rollback is necessary.

**Review by TomasTurina — APPROVED** [review]

**Review by TomasTurina — APPROVED** [review]

**Pull request merged**